### DISCUSSION TITLE
Hardware listing and info?

### DISCUSSION THREAD

**Discussion by james-wallace-ghub** [mamedev/discussions#16] [Fruit machine emulation]
Given our quest for accuracy with this sort of thing, I'm trying to track down the ancilliary components of these things so that we can at least emulate their behaviour according to the stock datasheets etc, or ideally, dump the MCUs where they exist.

I'll braindump what I know here when I'm near my parts list, but offhand does anyone know what the BFM Scorp 4 alpha (BFM BDA)was actually based on? The command codes are clearly a custom thing, as most of the behaviour matches the original starburst (BFM BD1), aside from the new character set and a few other enhancements.

Main reason I ask is that MFME in the officially released source has a character font for what looks to be that alpha, in a column-based encoding (so y=0,y=1 etc to make the 8x5). The closest part match for the Scorpion I can see made by Futaba is row-based, which seems inefficient to me, but if we're going to do it, do it right I guess.

Any help gratefully received

**Comment by james-wallace-ghub** [comment]
So, as far as I can tell the non BFM single strip alphas are parts derived from the Samsung 16l02da4 (which is a clone of MSM9202). That addresses its characters column by column, which makes more sense.

**Comment by Paul-Arnold** [comment]
Had a search through my stuff but can't find anything on the early displays. I doubt even BFG have much/anything either - discs with BD1 and BDA source were thrown when the factory shut.

The design of the most recent dot matrix displays for use with Scorpion6 was outsourced to Noritake-Itron. These used a Noritake-Itron glass - MN09608A.
There were two versions:
GU96X8M-K629C1 and GU96X8M-K657C2
The only difference between the two was K657C2 had RGB led backlighting.
ATMEGA8515 was used on the C1, I don't have a C2 to check but don't recall it being changed.

All devices I've got are code protected, don't know if there's a way of bypassing that without de-capping ?

99% working emulated versions of these exist which could be converted for MAME if needed.

**Reply by james-wallace-ghub** [reply]
Well, anything will help, with FME being so splintered its hard to keep track. Main reason I've come back to try to help with this is because so much of this seems to have been lost to posterity because of Chris' passing.

As for SC6, I don't think we do much with that, but I have some of the truth table already from some enterprising hardware hacker. (the C2 version)

**Comment by Paul-Arnold** [comment]
OK, I need to add a couple of the new commands to my emulation then I’ll convert and add to MAME.
Testing is a pain as there are new commands which have never been used by  games :(

SC6 doesn’t have a placeholder in MAME yet. At some point I’ll add what I can.
Still trying to get hold of some update files which are loaded via USB for the Reflex Games version of SC6. Don’t suppose you have any ?

**Reply by johnparker007** [reply]
Hi @Paul-Arnold  :) 

Would these be the Reflex SC6 update files you're after?

[PR2192 - Cash King S6.zip](https://github.com/mamedev/discussions/files/10232435/PR2192.-.Cash.King.S6.zip)
[PR2202 - Wild Dragons S6.zip](https://github.com/mamedev/discussions/files/10232436/PR2202.-.Wild.Dragons.S6.zip)
[Money Mad Hatter S6.zip](https://github.com/mamedev/discussions/files/10232438/Money.Mad.Hatter.S6.zip)

**Reply by Paul-Arnold** [reply]
Thanks very much for those John. That's exactly what I was looking for.
Have updated my board and they work :)
Just need to  work out the encryption now to save the hassle of dumping flash from the board.

Must get back onto the alpha at some point to but this is far more interesting.

Thanks again.

**Reply by SomeRandomGuyIdk** [reply]
Since you're mentioning flash dumps, is the game just stored there in plaintext? In that case the updater should be somewhere on that flash as well, together with keys for decrypting the updates 😄. Unless they're using the FPGA in some way for doing crypto, which complicates things a bit, but as long as it isn't some scheme where the FPGA sits between the flash and CPU and decrypts data on the fly, then it shouldn't be that hard to decrypt the games.

**Reply by Paul-Arnold** [reply]
Game is stored in flash in plaintext together with the bootloader/updater.
I had the bootloader and updater but had difficulty getting any of the Reflex update files to verify but John was kind enough to provide the above.
Nothing too special from Reflex.

Emulating a Coldfire at 240mhz needs a much faster pc than what I've got though.

BTW, Great work on the Black Box driver. Why are they still sitting on it ?

**Reply by SomeRandomGuyIdk** [reply]
So that's how Reflex did it at least, more of a security-through-obscurity thing but I guess they succeeded in keeping nosy people out 😄. I wonder if the updates are signed or not (not _that_ much point in doing it since a really determined person can just program the flash themselves to run custom code).

That's fair, though if there isn't any code in the games that makes CPU speed affect timings then you can probably get away with slashing the emulated CPU clock. Might end up needing hacks for peripheral baud rates etc. though. Makes more sense to get Scorpion 4/5 dealt with first though.

I'm sure Black Box will make it in eventually. I'm still learning the ropes here and Vas is finding dumb mistakes that got in, basing new things off legacy code also isn't helping 😄. At the end of the day it's a volunteer project, things take a while sometimes.

**Reply by Paul-Arnold** [reply]
There’s some sort of validation process which I’ll work out at some point but it’s only really needed if you want to load stuff via their loader.
Would be good to see SC4/5 progressed further and I’m surprised nobody has done them as there isn’t really anything tricky on them until you reach the point of emulating HI2 devices and that only applies to a relatively small number of games.

**Comment by Paul-Arnold** [comment]
Are there any working games supported by MAME which use the BFG dot matrix alphas ?

I would think updating the font is fairly easy but need something to test things like programable characters.

**Reply by james-wallace-ghub** [reply]
Funnily enough, no - but there are some that will get far enough to show you stuff, as Tafoid was pointing out that my code had 'regressed' since I was using the Scorp4 handlers for Scorp 5

sc5bsp will get at least as far as FATAL IIC 1 (or FATAL IIC 1G at the moment because of the windowing issue, but I'll see if I can work that out)
sc4dnd and the various DOND clones will also get somewhere I think that actually displays stuff.

**Reply by Paul-Arnold** [reply]
Thanks for the info. I'll take a look now I've finished with Reflex stuff.

IIRC the first FATAL IIC 1 is for the rtc/watchdog chip which isn't emulated yet. Not too difficult to add.
The next FATAL IIC 1 is for the protection chip. It's fairly easy to find in the code and not particularly complicated so could be simulated without too much trouble if somebody wanted to do it. Needs I2C on the cpu implementing which is more of a pain.

**Reply by Paul-Arnold** [reply]
The FATAL IIC1G is due to the clear command - 0xbX
In bfm_bd1 you've only got clear nothing and clear inside window, needs clear outside window and clear all putting back. 
bfm_bd1 looks ok but I didn't try that.

The flashing in bfm_bd1 looks a bit odd too but possibly I've not understood - didn't look too hard. If "flash inside window" is sent it should only flash whats inside the window at the point that command is received. If you then add a new character to the window the new char shouldn't flash but previously flashing chars should remain flashing. Same for "flash outside window".  Hope that makes sense ?

**Reply by james-wallace-ghub** [reply]
Main reason the flashing is bizarre is I only have the datasheet that was scanned ages ago, seemingly through a fax machine. While my image manipulation skills aren't bad, it was impossible to read that section. And then I couldn't find a workable example to test out.

Converting MFME style flashing code to our timing system was a challenge, to say the least.

BTW, how big is the font character list on the dot matrix alpha? I think the random lettering here is because of two things, one is I'm saying anything that doesn't have bit 7 set is a character, and then I'm ANDing with 0x3f to get an index. Probably on the real thing that's different, but I still have the starburst font in there. Again, MFME had some form of new font there, but I don't know how much I trust it, since it's stored by column it seems, and this code is clearly expecting to store characters in RAM rowwise.

As for Scorp 4 - I'm not a Bellfruit fan, and I was reluctant to do much to the SC4 stuff while that tech was still reasonably active in pubs through the myriad DOND knockoffs, Probably being too cautious, but still.

**Reply by Paul-Arnold** [reply]
I don't know what the early datasheets were like but the latest ones for the GU96X8M... are rubbish and don't explain the commands in any detail. For flashing they just say "D0-D3 Flash characters (none, window, background, all)."
The weird behaviour when sending multiple decimal points and other strangeness isn't explained at all.

The dot matrix font is 0x00-0x7f. In power-up default mode 0x00-0x3f is *almost* a mirror of 0x40-0x7f. Chars 0x25,0x26,0x3a,0x3b are blanks but 0x65 = percent, 0x66 = ampersand, 0x7a = colon and 0x7b = semicolon
It's also got a full upper/lower case mode which can be enabled.

Not sure how much Inspired Gaming is bothered about old tech, probably not much. As long as it's done without any prior knowledge is there much they could do other than ask you to stop ?

**Reply by james-wallace-ghub** [reply]
Probably not much, they haven't been fussed before, but MAME is higher profile, after all.

I don't really want to get in the way of your alpha work, but there may be some fixing needed to be done there anyway. I'll see if there's anything in the older techs for now just to make some progress, but I've been sidetracked with other projects that need some of my other skillset, and it's finding time to do both.

**Reply by Paul-Arnold** [reply]
I keep getting drawn to other stuff too, unfortunately stuff I can't release :(

I'll try and make a start on the alpha conversion this coming week or so. My font is stored as columns but whether that's right or not I don't know. I couldn't find any info on the actual glass used on the GU96X...

**Comment by Paul-Arnold** [comment]
Finally getting round to mame alpha stuff. Trying to get the dimming sorted so from other threads on here I've come up with this.
While it seems to work correctly I want to make sure this is the right approach.
I've assumed that treating the display as 8 rows of 96 instead of 96 columns of 8 is the correct thing to do even though I'm storing the font in columns ?

I'm not really sure what to do with the backlight but I don't think mame has anything which uses that anyway yet.

`
	<element name="vfd_dm_dot">
		<rect>
			<color state="0" red="0.0" green="0.075" blue="0.125" />
			<color state="1" red="0.0" green="0.6" blue="1.0" />
			<bounds left="0.0" top="0.0" right="1.0" bottom="1.0" />
		</rect>
	</element>

	<group name="vfd0">
		<repeat count="8">
			<param name="s" start="0" increment="96" />
			<param name="y" start="0" increment="1" />
			<repeat count="96">
				<param name="x" start="0" increment="1" />
				<element ref="vfd_dm_dot" >
					<animate name="vfdduty0" />
					<color state="8" red="0.0" green="1.0" blue="1.0" alpha="1.0" />
					<color state="0" red="0.0" green="1.0" blue="1.0" alpha="0.0" />
					<bounds x="~x~" y="~y~" width="0.8" height="0.8" />
				</element>
			</repeat>
		</repeat>
		<repeat count="8">
			<param name="s" start="0" increment="96" />
			<param name="y" start="0" increment="1" />
			<repeat count="96">
				<param name="n" start="~s~" increment="1" />
				<param name="x" start="0" increment="1" />
				<element name="vfddotmatrix~n~" ref="vfd_dm_dot" state="0">
					<animate name="vfdduty0" />
					<color state="0" red="0.0" green="1.0" blue="1.0" alpha="1.0" />
					<color state="8" red="0.0" green="1.0" blue="1.0" alpha="0.0" />
					<bounds x="~x~" y="~y~" width="0.8" height="0.8" />
				</element>
			</repeat>
		</repeat>
	</group>
`

**Reply by johnparker007** [reply]
Necroing an old dev thread here, but that does look like the alpha dimming strategy I did to work around the .lay system.  Not sure if it's 'right' but it does work, without introducing any further new functionality to the rendlay system, which I believe Cuavas is keen to avoid (as of course you can end up with never-ending edge cases).  

I believe in the absence of introducing a way to represent an off/idle/lit with normalised value between idle->lit in a vanilla mame .lay file, this is the way to go (at this present time), that is in keeping with minimising extra .lay functionality, at the expense of 'bloat' in the actual layout xml).

**Reply by Paul-Arnold** [reply]
Thanks John,
I kinda lost interest in the alpha as I’m not convinced the BFM stuff will be developed further by anyone anytime soon. There doesn’t seem to be much love for the Bell Fruit stuff, not sure why.

The above , or similar, is in the recent PR which has been tested on my Scorp6 emulation and works fine but the led backlight isn’t 100% right visually.

Would be nice to see BFM stuff working, especially the early video SWP but I just don’t know anyone who wants to get involved :(

**Reply by johnparker007** [reply]
I have done some very basic tweaks many moons ago to (fme) drivers, since worked on side projects such as Arcade Sim and the next version Oasis (that uses a native MAME executable as the emulation backend thru Lua)... hopefully x-platform (i.e: android) when finally getting somewhere.

I've been a bit under the weather for most of 2023, so very little coding from me at the mo, but thank you for your efforts so far, it's all much appreciated, as my plan is to use purely MAME for FME within the new iteration of AS (currently using MFME for FME).

Example of Arcade Sim (includes FME via MFME as proof-of-concept):
https://www.youtube.com/watch?v=dTWk4gIJCuM

Example of using vanilla MAME for FME (as proof-of-concept):
https://www.youtube.com/watch?v=S4rI25OkRJE

The second example is open source (first example is not able to be community-driven wrt new machines), and source is on github here: https://github.com/johnparker007/Oasis

Good luck with future MAME-based FME, hope to see more from you! :)

**Reply by Paul-Arnold** [reply]
Wow, amazing work. Very impressive. Wouldn’t have a clue how to even start something like that.
I’ll keep a lookout for updates

**Reply by johnparker007** [reply]
> Would be nice to see BFM stuff working, especially the early video SWP but I just don’t know anyone who wants to get involved :(

Many years ago (2016ish), a lot more of the Sc1/Sc2/Sc4 machines worked correctly.   Some Sc4 machines needed to be in Door Open mode, due to unemulated ccTalk system, though we did get a hacked in basic implementation working, but that code has been lost (just simulating the various packets really, nothing too crazy.  It was a guy Keith on the dadsfme forum who was looking at it, not heard from him).  

There's been various general refactors etc over the years that have led to these regressions.  So they may not take much work (relatively speaking of course), to fix those regressions and see some games become playable again.

I've had a nightmare year with health this year unfortunately, so I've not been up to doing much coding at all, and also I'm more focused on stuff that uses MAME as an emulation backend like those projects above - though, a part of this will be a layout editor (that will be able to export MAME .lay files and resources), so that should be helpful to MAME :)  So next year hoping to get back to a bit of FME dev here and there, health allowing.

> The above , or similar, is in the recent PR which has been tested on my Scorp6 emulation and works fine but the led backlight isn’t 100% right visually.

Great news - be nice to see a standard base .lay for sc6, maybe see a boot error message on the display for the one DOND game that is in there! :)  AFAIK no public emulator has ever ran any sc6 games.

**Reply by Paul-Arnold** [reply]
It wouldn’t require too much work to get SC5 a bit further with nvram and working sounds.
After that it requires I2C interface and last time I looked there wasn’t a I2C master driver or template in MAME and I don’t think Vas likes “instantaneous I2C”.
CCTalk isn’t too bad, I did DES hopper and coin validator for SC6.

SC6 in MAME was a major hack to the cpu core to try and get the speed anywhere near acceptable. Speed wouldn’t matter to get it to the point of a boot error message though.
Whether it’d be wise to make the decryption process public and risk upsetting Inspired Gaming is a different matter.

The biggest issue is I don’t find the MAME PR process enjoyable, actually I hate it. I don’t care that the format of a line in one part of the file is different to all the others or that there’s an extra space or a missing space etc. I kind of understand some of the reasons why they want consistency but it puts me off contributing :(

**Reply by johnparker007** [reply]
Cool, sc5 at least shouldn't be an issue wrt. manufacturer concerns, since it's already fully emulated in MFME (along with MPU5).

Unsure on sc6, I mean it is still a live platform to be fair driving most of the current gen EM fruit machines in pubs/arcades... though I don't recall there being any issues with MFME was updated to support sc5 - at that time, sc6 was still the current gen platform I think, or at least was a very popular platform for sited machines.  So probably not an issue, though you never know! :) 

Yeah - I can see how the code reviews/standards can get pretty tedious, but also necessary when you have such a behemoth of a codebase like MAME.  

I'm hoping to be a bit better next year, perhaps I could do the grunt work of the PR process, as I don't really mind doing that stuff!  My only slight problem with it is that I'm not a git expert, I've always used Perforce (now called Helix) for source control - it's much more common in the games industry, but of course it is not free like git!  I generally use GitKraken with a little command line as needed, does always seem to take me longer though, and is much more of a pain compared to Perforce.  Presumably I can do the equivalent of taking over one of your working pull requests, to then do the modifications as requested by Vas?

I'll post in here next year, if I start to get a bit better, and up to doing some coding again - as I really wouldn't mind if it means we get more emulation development in MAME.  While I can very slowly develop _some_ of the emulation stuff, it's not really my strong suit at all... but a bit of refactoring and clean up based on Vas's feedback is a breeze really.  So perhaps that could be something to make it more enjoyable for you again, and get some progress :)  

I have a vested interest, since my plan is to make my next version of the Arcade Simulator in Oasis, along with associated layout editing/testing tools etc all use standard MAME as the core backend for FME (instead of MFME).

**Reply by Paul-Arnold** [reply]
I really wouldn’t expect someone else to do the stuff I don’t like but it’s good of you to offer.

I hope you’re feeling better next year and can develop your projects further.

on the subject of SC6 you very kindly shared some Reflex update files.  I wondered if you had any for other games by other manufactures using reflex sc6 ?  I believe Blueprint did a number of games on the reflex board.

**Reply by johnparker007** [reply]
Haha no worries, it's a genuine offer though, cleaning up C code is pretty relaxed work (apart from dealing with git) :)  Hoping to be on the mend soon :) 

I have put a request out on a couple of forums, to see about sourcing more of those Reflex-based sc6 update files - I'll post back in here with any additional Update files I can procure...

**Reply by johnparker007** [reply]
@Paul-Arnold I have procured a few more SC6 update file sets.  Some are Reflex, some are just labelled S6.  Unfortunately these seem a little hard to come by, compared to the SC5 and below ROMs, probably because those older ROMs could be used in the existing MFME emulator, so were more useful to the scene.

**Reply by johnparker007** [reply]
I have also just been passed a mixed zip containing more mainly SC6 software - unsure which files will be of use to you @Paul-Arnold.

**Reply by Osso13** [reply]
Hi, could you please get in touch directly? We'd like to avoid links to possibly copyrighted stuff in here. Thanks!
I'll remove the links.

**Reply by johnparker007** [reply]
Sorry @Osso13 - no probs will DM Paul, ta :) 

@Paul-Arnold my email address is _gyrox007 at gmail dot com_, if you send me an email from your address I can directly send you the SC6 resources (I don't see a way to send DM via github).

**Reply by Paul-Arnold** [reply]
Sorry @Osso13 for any problems this caused.
I've emailed John.

**Reply by cuavas** [reply]
You guys really seem to want to take it out on me, while also complaining that these fruit machines are prone to regressing.  Have you considered that the whole reason these systems tend to suffer more regressions than average is because the code is a huge mess, and hence difficult to understand and maintain?  A major reason for having standards it to help keep the code maintainable and avoid having things regress in the first place.

Think about this from my point of view.  A lot of pull request feedback is the same thing over and over again, often with the same people.  An incredibly frequent thing I catch is not guarding against debugger accesses in read handlers with side effects.  Do you think I enjoy explaining why it’s important hundreds of times?  In many cases, I catch stuff that *is* broken before it gets in.

If we went with an “anything goes” approach, we’re just going to end up with a big unmaintainable mess, and regressions everywhere.  I already have trouble getting bigger pieces of work on MAME done because of the amount of time I have to spend fighting fires.  I don’t need even more of that on my plate.

Maintaining consistent formatting for your code is a pretty small thing to ask.  Most projects have far stricter coding standards than MAME.

**Reply by johnparker007** [reply]
I'm sorry @cuavas if anything I posted came across as complaining, or having a go at you - it was honestly not my intention at all!

I only bought up the regressions to let Paul know that the machines had worked in the past, so most/all the code was complete, but with bugs (rather than the emulation having never worked, which might suggest additional reverse-engineering was still needed).

I really do appreciate the tough but necessary work of code reviewing that you do, to keep the massive MAME codebase in order, and not descending into chaos! :)

**Reply by Paul-Arnold** [reply]
Hi @cuavas I wasn’t trying to have a go at you so I’m sorry it came across that way.
I do understand the reasons for maintaining standards and appreciate it takes a lot of your time checking PRs. I certainly wouldn’t want the standards to drop. I was merely pointing out it’s something that puts me off contributing but that’s my problem and not yours.
For whatever reason inconsistent formatting is something I find really difficult to spot so I find it frustrating when I think there isn’t any and the PR process shows there are.
Maybe it’s too much reliance on code formatters for my other projects.

I will try and do better in the future and not let it annoy me so much.

**Reply by Paul-Arnold** [reply]
I've now had chance to look at the Reflex update files. I've 105 update files but some are different releases of the same game, then there's des/non-des and protocol/non-protocol.
Games are by Reflex and Betcom, I'm fairly sure INNOV8 used that tech too, I'll ask next time I bump into someone from there.
Interesting that Betcom added additional security in the form of an I2C device which acts as both security and eeprom whereas Reflex didn't have anything.

I might have a look to see what's regressed on the old tech, I've no idea what used to work but doesn't now though.
Will also add a bit to the SC5 memory map and add nvram and sounds so it at least errors correctly.
After that it needs an I2C master implementing as does Scorp4.  Not sure whether "instantaneous  i2c" would be accepted now even though lots of drivers use it.

**Reply by johnparker007** [reply]
That all sounds good man :)  

In terms of machines that worked back around 2016 but now do not, I know a common new issue on the Scorpion 4 machines is GEN METER ERR 60 during boot.

![image](https://github.com/mamedev/discussions/assets/17624161/4187f073-eb40-416e-8a10-1d7e51bce0ac)

The machine pictured is the original Deal Or No Deal (sc4dnd) - but this issue has affected quite a lot of the sc4 machines (not all though).  Might tie in nicely since you are working on sc4 and sc5 anyway - good luck :)

**Reply by Paul-Arnold** [reply]
The meter error problem was introduced in mame0213 in September 2019. 
The Starpoint Electronic Counter (SEC) was converted to a proper MAME device.
When writing the data bit the old version would do m_data = bit ? 1 : 0;
The new version doesn't convert to 1 or 0 and just copies the value to m_data so only works correctly if the caller is passing 0 or 1. Sc4 passes 0 or 0x10.
Re-instating the 1:0 conversion fixes the issue :)
I'll do a PR tomorrow. Ooops, later today :)

**Reply by rb6502** [reply]
I'd recommend instead fixing sc4 to pass 0 or 1, because that's likely more correct to how it's hooked up.  BIT(whatever, 4) should do the trick.

**Reply by johnparker007** [reply]
Awesome work :)  I'll download a ~2016 version and see if I can find some more of those machines that have developed issues since then, that Gen Meter Error sc4 one affected lots of sc4 machines though, a good fix!

**Reply by Paul-Arnold** [reply]
@rb6502   Should the clk_w function in sec.cpp be the same then ?
ie, remove the current state = state ? 1:0 and force drivers to pass 1 or 0 ?
That would make the two functions consistent.

**Reply by rb6502** [reply]
@Paul-Arnold Yes, that would be great.

**Comment by james-wallace-ghub** [comment]
An actual layout editor would be great - the main reason I've pulled back
from this has been that we need better art support before we can
realistically do anything, and I just totally burned out.

Any obvious gaps in the 'old' techs I should probably try to probe as I
think a lot of the 6800 series peripheral space is different now than then,
and regressions may be rife.

**Reply by johnparker007** [reply]
I had been working on doing bulk conversions I think around end of last year, but then got ill.  New approach is a project called 'Oasis', that will be able extract and import MFME layouts to its own internal layout format (just simple JSON).  So when I get back to this, I need to continue with that, and then implement 'text' components, so the 'classic' layouts can be processed from MFME and ultimately exported as .lay files.

Later I'll be adding in editing facilities directly in the Oasis editor... currently, it is of course depending on layouts built in MFME.

Will take longer but I think it will be the way to go, so then others can sort out all the MAME internal layouts using Oasis without needing me to do anything :)
